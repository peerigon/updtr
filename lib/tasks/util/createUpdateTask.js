"use strict";

function createUpdateTask(instance, outdated) {
    return {
        name: outdated.name,
        type: outdated.type,
        updateTo: instance.wanted ? outdated.wanted : outdated.latest,
        rollbackTo: outdated.current,
    };
}

module.exports = createUpdateTask;
