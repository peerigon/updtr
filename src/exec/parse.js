const STRING_PROPERTIES = ["name", "current", "wanted", "latest"];

/**
 * Applies JSON.parse on the str.
 * Additionally, returns null if the str was empty or just contained whitespace.
 *
 * @param {string} str
 * @returns {*}
 */
function parse(str) {
    const trimmed = str.trim();

    if (trimmed.length === 0) {
        return null;
    }

    let resultJSON = {};
    try {
        resultJSON = JSON.parse(trimmed);
    } catch (e) {
        const eol = require("os").EOL;
        const lines = trimmed.split(eol);
        lines.map(stringJSON => {
            const possible = JSON.parse(stringJSON);
            if (possible.type === "table") {
                resultJSON = possible;
            }
        })
    }

    return resultJSON;
}

function returnIfValid(result) {
    STRING_PROPERTIES.forEach(prop => {
        if (typeof result[prop] !== "string") {
            throw new Error("Unexpected output format of package manager");
        }
    });

    return result;
}

function tryParse(normalizer) {
    return (stdout, cmd) => {
        try {
            return normalizer(parse(stdout));
        } catch (err) {
            err.message = `Error when trying to parse stdout from command '${ cmd }': ${ err.message }`;
            throw err;
        }
    };
}

function arrToObj(arr, keys) {
    return keys.reduce((obj, key, i) => {
        obj[key] = arr[i];

        return obj;
    }, {});
}

// By sorting the parsed data, we get deterministic results across different npm and yarn versions.
// As a nice side-effect, a package like eslint will always be updated before eslint-config-peerigon
// which might have a peer dependency on eslint
// See https://github.com/peerigon/updtr/issues/48
function sortByName(o1, o2) {
    return o1.name > o2.name;
}

const list = tryParse(
    parsed =>
        (parsed.dependencies === undefined ?
            [] :
            Object.keys(parsed.dependencies)
                .map(name => ({
                    name,
                    version: parsed.dependencies[name].version,
                }))
                .sort(sortByName))
);

export default {
    npm: {
        outdated: tryParse(parsed => {
            if (parsed === null) {
                return [];
            }

            const names = Object.keys(parsed);

            return names
                .map(name => parsed[name])
                .map((dep, index) =>
                    returnIfValid({
                        name: names[index],
                        current: dep.current,
                        wanted: dep.wanted,
                        latest: dep.latest,
                    })
                )
                .sort(sortByName);
        }),
        list,
    },
    yarn: {
        outdated: tryParse(
            parsed =>
                (parsed === null ?
                    [] :
                    parsed.data.body
                        .map(row => arrToObj(row, parsed.data.head))
                        .map(dep =>
                            returnIfValid({
                                name: dep.Package,
                                current: dep.Current,
                                wanted: dep.Wanted,
                                latest: dep.Latest,
                            })
                        )
                        .sort(sortByName))
        ),
        // We currently only use npm for the list command (see cmds.js).
        // Put the real implementation here if we decide to use yarn for list.
        list,
    },
};
