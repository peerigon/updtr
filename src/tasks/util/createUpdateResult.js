export default function createUpdateResult(updateTask, success) {
    return {
        name: updateTask.name,
        updateTo: updateTask.updateTo,
        rollbackTo: updateTask.rollbackTo,
        success,
    };
}
