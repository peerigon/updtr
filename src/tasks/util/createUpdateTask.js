import {
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
} from "../../constants/config";

function determineUpdateToVersion({ current, wanted, latest }, { updateTo }) {
    switch (updateTo) {
        case UPDATE_TO_LATEST:
            return latest;
        case UPDATE_TO_WANTED:
            return wanted;
        case UPDATE_TO_NON_BREAKING:
        default:
            return "^" + current;
    }
}

export function isUpdateToNonBreaking(updateTask) {
    return updateTask.updateTo === "^" + updateTask.rollbackTo;
}

export default function createUpdateTask(outdated, updtrConfig) {
    return {
        name: outdated.name,
        updateTo: determineUpdateToVersion(outdated, updtrConfig),
        rollbackTo: outdated.current,
    };
}
