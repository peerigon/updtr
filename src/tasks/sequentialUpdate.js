import Sequence from "./util/Sequence";
import createUpdateResult from "./util/createUpdateResult";
import updateTo from "./util/updateTo";
import rollbackTo from "./util/rollbackTo";

function renderUpdate(updtr, updateTask) {
    return updtr.cmds.install({
        registry: updtr.config.registry,
        modules: [updateTo(updateTask)],
    });
}

function renderTest(updtr) {
    return updtr.cmds.test();
}

function renderRollback(updtr, failedUpdateTask, nextUpdateTask) {
    const modules = [rollbackTo(failedUpdateTask)];

    if (nextUpdateTask !== undefined) {
        modules.push(updateTo(nextUpdateTask));
    }

    return updtr.cmds.install({
        registry: updtr.config.registry,
        modules,
    });
}

async function runUpdateTask(sequence, updateTasks, i, previousUpdateResults) {
    const updateResults = await previousUpdateResults;
    const previousUpdateResult = updateResults[updateResults.length - 1];
    const updateTask = updateTasks[i];
    // If the previous update was a failure, we don't need to update now because
    // during the rollback, the next update is also installed in parallel
    const updateNecessary = previousUpdateResult === undefined ?
        true :
        previousUpdateResult.success === true;

    sequence.baseEvent = {
        updateTasks: {
            current: i + 1,
            total: updateTasks.length,
        },
        ...updateTask,
    };
    const updtr = sequence.updtr;
    let testResult;
    let success;

    if (updateNecessary === true) {
        await sequence.exec("updating", renderUpdate(updtr, updateTask));
    }

    try {
        testResult = await sequence.exec("testing", renderTest(updtr));
        success = true;
    } catch (err) {
        // Remember: instanceof Error might not work in Jest as expected
        // https://github.com/facebook/jest/issues/2549
        testResult = err;
        success = false;
    }

    sequence.baseEvent.success = success;

    if (success === false) {
        const nextUpdateTask = i + 1 < updateTasks.length ?
            updateTasks[i + 1] :
            undefined;

        await sequence.exec(
            "rollback",
            renderRollback(updtr, updateTask, nextUpdateTask)
        );
    }

    sequence.emit("result", {
        stdout: testResult.stdout,
    });

    return updateResults.concat(createUpdateResult(updateTask, success));
}

export default (async function sequentialUpdate(
    updtr,
    updateTasks,
    previousUpdateResult
) {
    const sequence = new Sequence("sequential-update", updtr, {
        updateTasks,
    });

    if (updateTasks.length === 0) {
        return [];
    }

    sequence.start();

    const updateResults = await updateTasks.reduce(
        (updateResults, updateTask, i) =>
            runUpdateTask(sequence, updateTasks, i, updateResults),
        previousUpdateResult === undefined ? [] : [previousUpdateResult]
    );

    if (previousUpdateResult !== undefined) {
        // The previousUpdateResult is the first element in the updateResults array, so let's remove it.
        updateResults.shift();
    }

    sequence.baseEvent = {
        updateResults,
    };

    sequence.end();

    return updateResults;
});
