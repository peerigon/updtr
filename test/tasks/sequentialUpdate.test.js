import sequentialUpdate from "../../src/tasks/sequentialUpdate";
import FakeUpdtr from "../helpers/FakeUpdtr";
import createUpdateTask from "../../src/tasks/util/createUpdateTask";
import createUpdateResult from "../../src/tasks/util/createUpdateResult";
import readFixtures from "../helpers/readFixtures";
import parse from "../../src/exec/parse";
import pickEvents from "../helpers/pickEvents";
import {
    execError,
    update,
    testPass,
    testFailWithRollback,
    errorExecUpdate,
    errorExecRollback,
} from "../fixtures/execResults";

let stdoutLogs;

function createUpdateTasks(updtrConfig) {
    const packageManager = updtrConfig.use;
    const outdated = parse[packageManager].outdated(
        stdoutLogs.get(`outdated/outdated.${ packageManager }.log`)
    );

    return outdated.map(outdated => createUpdateTask(outdated, updtrConfig));
}

beforeAll(async () => {
    stdoutLogs = await readFixtures([
        "outdated/outdated.npm.log",
        "outdated/outdated.yarn.log",
    ]);
});

describe("sequentialUpdate()", () => {
    describe("when the given updateTasks array is empty", () => {
        it("should resolve immediately with an empty array without emitting events", async () => {
            const updtr = new FakeUpdtr();

            expect(await sequentialUpdate(updtr, [])).toEqual([]);
            expect(updtr.exec.args).toEqual([]);
            expect(updtr.emit.args).toEqual([]);
        });
    });
    describe("when the given updateTasks array contains update tasks", () => {
        describe("using npm", () => {
            describe("when the tests succeed", () => {
                it("should return the expected update results", async () => {
                    const updtr = new FakeUpdtr();
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(
                        testPass,
                        update,
                        testPass
                    );

                    const updateResults = await sequentialUpdate(
                        updtr,
                        updateTasks
                    );

                    expect(updateResults).toMatchSnapshot(
                        "updateResults success"
                    );
                });
            });
            describe("when the first test fails and the rest succeeds", () => {
                it("should emit expected events and execute expected commands", async () => {
                    const updtr = new FakeUpdtr();
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(
                        testFailWithRollback,
                        testPass
                    );

                    const updateResults = await sequentialUpdate(
                        updtr,
                        updateTasks
                    );

                    expect(updtr.exec.args).toMatchSnapshot("exec.args npm");
                    expect(updtr.emit.args).toMatchSnapshot("emit.args npm");
                    expect(updateResults).toMatchSnapshot(
                        "updateResults one fail rest pass"
                    );
                });
            });
        });
        describe("using yarn", () => {
            describe("when the tests succeed", () => {
                it("should return the expected update results", async () => {
                    const updtr = new FakeUpdtr({
                        use: "yarn",
                    });
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(
                        testPass,
                        update,
                        testPass
                    );

                    const updateResults = await sequentialUpdate(
                        updtr,
                        updateTasks
                    );

                    expect(updateResults).toMatchSnapshot(
                        "updateResults success"
                    );
                });
            });
            describe("when the first test fails and the rest succeeds", () => {
                it("should emit expected events and execute expected commands", async () => {
                    const updtr = new FakeUpdtr({
                        use: "yarn",
                    });
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(
                        testFailWithRollback,
                        testPass
                    );

                    const updateResults = await sequentialUpdate(
                        updtr,
                        updateTasks
                    );

                    expect(updtr.exec.args).toMatchSnapshot("exec.args yarn");
                    expect(updtr.emit.args).toMatchSnapshot("emit.args yarn");
                    expect(updateResults).toMatchSnapshot(
                        "updateResults one fail rest pass"
                    );
                });
            });
        });
    });
    describe("when there is a previous update result", () => {
        it("should skip the first updating step if the previous update success was false", async () => {
            const updtr = new FakeUpdtr();
            const updateTasks = createUpdateTasks(updtr.config);
            const previousUpdateResult = createUpdateResult(
                updateTasks[0],
                false
            );

            updtr.execResults = testPass.concat(update, testPass);
            await sequentialUpdate(updtr, updateTasks, previousUpdateResult);

            expect(
                // first one is sequence start
                updtr.emit.secondCall.calledWith("sequential-update/testing")
            ).toBe(true);
            expect(
                pickEvents("sequential-update/updating", updtr.emit.args)
            ).toHaveLength(1);
        });
        it("should not skip the first updating step if the previous update success was true", async () => {
            const updtr = new FakeUpdtr();
            const updateTasks = createUpdateTasks(updtr.config);
            const previousUpdateResult = createUpdateResult(
                updateTasks[0],
                true
            );

            updtr.execResults = update.concat(testPass, update, testPass);
            await sequentialUpdate(updtr, updateTasks, previousUpdateResult);

            expect(
                // first one is sequence start
                updtr.emit.secondCall.calledWith("sequential-update/updating")
            ).toBe(true);
            expect(
                pickEvents("sequential-update/updating", updtr.emit.args)
            ).toHaveLength(2);
        });
        it("should not include the previous update result in the returned results", async () => {
            const updtr = new FakeUpdtr();
            const updateTasks = createUpdateTasks(updtr.config);
            const previousUpdateResult = createUpdateResult(
                updateTasks[0],
                false
            );

            updtr.execResults = testPass.concat(update, testPass);
            const results = await sequentialUpdate(
                updtr,
                updateTasks,
                previousUpdateResult
            );

            expect(results).toHaveLength(updateTasks.length);
        });
    });
    describe("unexpected errors", () => {
        it("should completely bail out if the update cmd exits with a non-zero exit code", async () => {
            const updtr = new FakeUpdtr();
            const updateTasks = createUpdateTasks(updtr.config);
            let givenErr;

            updtr.execResults = errorExecUpdate;

            try {
                await sequentialUpdate(updtr, updateTasks);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBe(execError);
            // emitted events: start, updating
            expect(updtr.emit.args.length).toBe(2);
        });
        it("should completely bail out if the rollback cmd exits with a non-zero exit code", async () => {
            const updtr = new FakeUpdtr();
            const updateTasks = createUpdateTasks(updtr.config);
            let givenErr;

            updtr.execResults = errorExecRollback;

            try {
                await sequentialUpdate(updtr, updateTasks);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBe(execError);
            // emitted events: start, updating, testing, rollback
            expect(updtr.emit.args.length).toBe(4);
        });
    });
});
