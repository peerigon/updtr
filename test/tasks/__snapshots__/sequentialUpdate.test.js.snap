// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emit.args npm 1`] = `
Array [
  Array [
    "sequential-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "sequential-update/updating",
    Object {
      "cmd": "npm install \\"updtr-test-module-1@2.0.0\\"",
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/testing",
    Object {
      "cmd": "npm test",
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/rollback",
    Object {
      "cmd": "npm install \\"updtr-test-module-1@1.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "success": false,
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/result",
    Object {
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "stdout": "Oh noez",
      "success": false,
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/testing",
    Object {
      "cmd": "npm test",
      "name": "updtr-test-module-2",
      "rollbackTo": "2.0.0",
      "updateTasks": Object {
        "current": 2,
        "total": 2,
      },
      "updateTo": "2.1.1",
    },
  ],
  Array [
    "sequential-update/result",
    Object {
      "name": "updtr-test-module-2",
      "rollbackTo": "2.0.0",
      "stdout": "Everything ok",
      "success": true,
      "updateTasks": Object {
        "current": 2,
        "total": 2,
      },
      "updateTo": "2.1.1",
    },
  ],
  Array [
    "sequential-update/end",
    Object {
      "updateResults": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": false,
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
]
`;

exports[`emit.args yarn 1`] = `
Array [
  Array [
    "sequential-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "sequential-update/updating",
    Object {
      "cmd": "yarn add \\"updtr-test-module-1@2.0.0\\"",
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/testing",
    Object {
      "cmd": "yarn test",
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/rollback",
    Object {
      "cmd": "yarn add \\"updtr-test-module-1@1.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "success": false,
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/result",
    Object {
      "name": "updtr-test-module-1",
      "rollbackTo": "1.0.0",
      "stdout": "Oh noez",
      "success": false,
      "updateTasks": Object {
        "current": 1,
        "total": 2,
      },
      "updateTo": "2.0.0",
    },
  ],
  Array [
    "sequential-update/testing",
    Object {
      "cmd": "yarn test",
      "name": "updtr-test-module-2",
      "rollbackTo": "2.0.0",
      "updateTasks": Object {
        "current": 2,
        "total": 2,
      },
      "updateTo": "2.1.1",
    },
  ],
  Array [
    "sequential-update/result",
    Object {
      "name": "updtr-test-module-2",
      "rollbackTo": "2.0.0",
      "stdout": "Everything ok",
      "success": true,
      "updateTasks": Object {
        "current": 2,
        "total": 2,
      },
      "updateTo": "2.1.1",
    },
  ],
  Array [
    "sequential-update/end",
    Object {
      "updateResults": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": false,
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
]
`;

exports[`exec.args npm 1`] = `
Array [
  Array [
    "npm install \\"updtr-test-module-1@2.0.0\\"",
  ],
  Array [
    "npm test",
  ],
  Array [
    "npm install \\"updtr-test-module-1@1.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
  ],
  Array [
    "npm test",
  ],
]
`;

exports[`exec.args yarn 1`] = `
Array [
  Array [
    "yarn add \\"updtr-test-module-1@2.0.0\\"",
  ],
  Array [
    "yarn test",
  ],
  Array [
    "yarn add \\"updtr-test-module-1@1.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
  ],
  Array [
    "yarn test",
  ],
]
`;

exports[`updateResults one fail rest pass 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": false,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
]
`;

exports[`updateResults one fail rest pass 2`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": false,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
]
`;

exports[`updateResults success 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
]
`;

exports[`updateResults success 2`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
]
`;
